02-peer-protocol.md

```diff
@@ -390,6 +400,377 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+  * [`4`:`minimum_depth`]
+  * [`2`:`to_self_delay`]
+  * [`2`:`max_accepted_htlcs`]
+  * [`33`:`funding_pubkey`]
+  * [`33`:`revocation_basepoint`]
+  * [`33`:`payment_basepoint`]
+  * [`33`:`delayed_payment_basepoint`]
+  * [`33`:`htlc_basepoint`]
+  * [`33`:`first_per_commitment_point`]
+  * [`2`:`opening_tlv_len`]
+  * [`opening_tlv_len`:`opening_tlv`]
+
+Rationale and Requirements are the same as listed above,
+for [`accept_channel`](#the-accept_channel-message)
+
+### The `funding_compose` Message
+
+This message exchanges the transaction input and output
+information necessary to compose the funding transaction.
+
+1. type: 58 (`funding_compose`)
+2. data:
+  * [`32`:`temporary_channel_id`]
+  * [`2`:`num_inputs`]
+  * [`num_inputs`:`input_info`]
+  * [`2`:`num_outputs`]
+  * [`num_outputs`:`output_info`]
+
+1. subtype `input_info`
+2. data:
+  * [`8`:`satoshis`]
+  * [`32`:`prevtxid`]
+  * [`4`:`prevtxoutnum`]
+  * [`2`:`max_witness_len`]
+  * [`2`:`scriptlen`]
+  * [`scriptlen`:`script`]
+
+1. subtype `output_info`
+2. data:
+  * [`8`:`satoshis`]
+  * [`2`:`scriptlen`]
+  * [`scriptlen`:`script`]
+
+#### Requirements
+
+The sending node:
+  - MUST ensure each `input_info` refers to an existing UTXO
+  - MUST ensure the `output_info`.`script` is a standard script
+  - MUST NOT spend the UTXOs it has specified (until/unless
+    channel establishment fails).
+  - MUST NOT include the channel funding output.
+  - if is the `opener`:
+    - MUST NOT send zero inputs (`num_inputs` cannot be zero).
+    - MUST specify an output with value zero, which will be used
+      as the change address.
+  - if is the `accepter`:
+    - consider the `put_limit` the total of `num_inputs` plus
+      `num_outputs` from `funding_compose`, with minimum 2.
+    - MUST NOT send a number of `input_info` or `output_info` which
+      exceeds the `put_limit`.
+    - MAY send zero inputs and/or outputs.
+
+The receiving node:
+  - if is the `opener`:
+    - MUST fail the channel if:
+      - the sum of `input_info` and `output_info` is greater than
+        the `put_limit`.
+  - otherwise:
+    - MUST fail the channel if:
+      - `change_satoshis` would be negative.
+  - if has not yet sent a `funding_compose`:
+    - MUST send their `funding_compose` message.
+  - otherwise:
+    - MUST use the sent and received `input_info` and `output_info`
+      to create the funding transaction, using `max_extra_witness_len`
+      for each `input_info` and `feerate_per_kw_funding` as specified in
+      `open_channel2`.
+    - MUST send `commitment_signed2`.
+
+
+#### Rationale
+Each node must have a complete set of the transaction inputs and outputs,
+to derive the funding transaction hash. This avoids information
+asymmetry between the nodes, as both sides share their input utxo set.
+
+`satoshis` is the value of the input or output. The opener must provide
+one change address for the transaction, which is flagged with a
+`satoshis` value of zero.
```

Comment:
You said this below as well; delete one?  We try to avoid spec-sounding language (i.e. "must") in Rationales to avoid confusion; I would say "The opener provides one output with `satoshis` equal zero, as a change address once the amount of change can be calculated".